CoreScriptsRoot/Modules/VR/LaserPointer.lua

```diff
@@ -0,0 +1,548 @@
+
+	SUCCESS_SOUND = "rbxassetid://147722227",
+	FAIL_SOUND = "rbxassetid://138087015",
+
+	PATH_RECOMPUTE_DIST_THRESHOLD = 4
+}
+
+local LASER = {
+	ARC_COLOR_GOOD = TELEPORT.ARC_COLOR_GOOD,
+	ARC_COLOR_BAD = TELEPORT.ARC_COLOR_BAD,
+	ARC_THICKNESS = 0.025,
+
+	MAX_DISTANCE = 100
+}
+
+local zeroVector2, identityVector2 = Vector2.new(0, 0), Vector2.new(1, 1)
+local zeroVector3, identityVector3 = Vector3.new(0, 0, 0), Vector3.new(1, 1, 1)
+local flattenMask = Vector3.new(1, 0, 1) --flattens a direction vector when multiplied by removing the vertical component
+local minimumPartSize = Vector3.new(0.2, 0.2, 0.2)
+local identity = CFrame.new()
+
+local function applyExpCurve(x, exp)
+	local y = x ^ exp
+	if y ~= y then
+		y = math.abs(x) ^ exp
+	end
+	return y
+end
+
+local LaserPointer = {}
+LaserPointer.__index = LaserPointer
+
+function LaserPointer.new()
+	local self = setmetatable({}, LaserPointer)
+
+	self.enabled = false
+
+	self.teleportMode = true
+	self.teleportMaxRangePerSecond = 16
+	self.teleportRangeT = 0
+	self.teleportPoint = zeroVector3
+	self.teleportNormal = Vector3.new(0, 1, 0)
+	self.teleportPart = nil
+	self.teleportValid = false
+	self.teleportButtonDown = false
+	self.teleporting = false
+	self.teleportBounceStart = tick()
+
+	self.pathValid = false
+	self.computingPath = false
+	self.pathStart = zeroVector3
+	self.pathEnd = zeroVector3
+
+	do --Create the instances that make up the Laser Pointer
+		--Create the ParabolaAdornment first; if the class doesn't exist, we'll exit early.
+		if not pcall(function()
+			self.parabola = Utility:Create("ParabolaAdornment") {
+				Name = "LaserPointerParabola",
+				Parent = CoreGui,
+				A = -1,
+				B = 2,
+				C = 0,
+				Color3 = TELEPORT.COLOR_GOOD,
+				Thickness = TELEPORT.ARC_THICKNESS
+			}
+			end) then
+			return nil
+		end
+
+		self.originPart = Utility:Create("Part") {
+			Name = "LaserPointerOrigin",
+			Anchored = true,
+			CanCollide = false,
+			TopSurface = Enum.SurfaceType.SmoothNoOutlines,
+			BottomSurface = Enum.SurfaceType.SmoothNoOutlines,
+			Material = Enum.Material.SmoothPlastic,
+			Size = minimumPartSize,
+			Transparency = 1 --smallest size possible
+		}
+		self.parabola.Adornee = self.originPart
+		
+
+		self.plopPart = Utility:Create("Part") {
+			Name = "LaserPointerTeleportPlop",
+			Anchored = true,
+			CanCollide = false,
+			Size = minimumPartSize,
+			Transparency = 1
+		}
+		self.plopBall = Utility:Create("Part") {
+			Name = "LaserPointerTeleportPlopBall",
+			Anchored = true,
+			CanCollide = false,
+			TopSurface = Enum.SurfaceType.SmoothNoOutlines,
+			BottomSurface = Enum.SurfaceType.SmoothNoOutlines,
+			Material = Enum.Material.Neon,
+			BrickColor = TELEPORT.PLOP_BALL_COLOR_GOOD,
+			Shape = Enum.PartType.Ball,
+			Size = identityVector3 * TELEPORT.PLOP_BALL_SIZE
+		}
+		self.plopAdorn = Utility:Create("ImageHandleAdornment") { --this feels hacky, but no good alternatives for unlit image in 3D that aren't SurfaceGuis w/ ImageLabels
```

Comment:
Since I didn't answer your question, they're an adornment that renders an image on a flat plane a bit like a SurfaceGui, but it has its own CFrame property and renders the image itself rather than through an ImageLabel. Lot less overhead that way.